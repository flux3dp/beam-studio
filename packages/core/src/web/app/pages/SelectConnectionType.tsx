import classNames from 'classnames';
import React, { useMemo } from 'react';

import useI18n from 'helpers/useI18n';
import { supportUsbModels } from 'app/actions/beambox/constant';

import { WorkAreaModel } from 'app/constants/workarea-constants';
import { useLocation } from 'react-router-dom';
import styles from './SelectConnectionType.module.scss';

const TYPE_URL_MAP = {
  wifi: '#initialize/connect/connect-wi-fi',
  wired: '#initialize/connect/connect-wired',
  ether2ether: '#initialize/connect/connect-ethernet',
  usb: '#initialize/connect/connect-usb',
};

const SelectConnectionType = (): JSX.Element => {
  const lang = useI18n().initialize;
  const { search } = useLocation();
  const model = useMemo(() => new URLSearchParams(search).get('model') as WorkAreaModel, [search]);

  const handleBack = () => {
    window.location.hash = '#initialize/connect/select-machine-model';
  };

  const handleConnectionTypeClick = (type: 'wifi' | 'wired' | 'ether2ether' | 'usb') => {
    const url = TYPE_URL_MAP[type];
    const urlParams = new URLSearchParams({ model });
    const queryString = urlParams.toString();

    window.location.hash = `${url}?${queryString}`;
  };

  const renderConnectionTypeButton = (
    type: 'wifi' | 'wired' | 'ether2ether' | 'usb'
  ): JSX.Element => (
    <button
      id={`connect-${type}`}
      type="button"
      className={classNames('btn', styles.btn)}
      onClick={() => handleConnectionTypeClick(type)}
    >
      {lang.connection_types[type]}
    </button>
  );

  const renderConnectionTypeContainer = (
    type: 'wifi' | 'wired' | 'ether2ether' | 'usb'
  ): JSX.Element => (
    <div className={styles['btn-container']}>
      <img className={styles.icon} src={`img/init-panel/icon-${type}.svg`} draggable="false" />
      {renderConnectionTypeButton(type)}
    </div>
  );

  return (
    <div className={styles.container}>
      <div className={styles['top-bar']} />
      <div className={styles.btns}>
        <div className={classNames(styles.btn, styles.primary)} onClick={handleBack}>
          {lang.back}
        </div>
      </div>
      <div className={styles.main}>
        <h1 className={styles.title}>{lang.select_connection_type}</h1>
        <div className={styles.btns}>
          {renderConnectionTypeContainer('wifi')}
          {renderConnectionTypeContainer('wired')}
          {renderConnectionTypeContainer('ether2ether')}
          {supportUsbModels.has(model) && renderConnectionTypeContainer('usb')}
        </div>
      </div>
    </div>
  );
};

export default SelectConnectionType;
