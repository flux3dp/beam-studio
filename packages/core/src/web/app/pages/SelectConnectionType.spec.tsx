import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import SelectConnectionType from './SelectConnectionType';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '#/initialize/connect/select-connection-type',
    search: '?model=ado1',
  }),
}));

const mockShowLoadingWindow = jest.fn();
jest.mock('app/actions/dialog-caller', () => ({
  showLoadingWindow: (...args) => mockShowLoadingWindow(...args),
}));

jest.mock('helpers/useI18n', () => () => ({
  initialize: {
    select_connection_type: 'How do you wish to connect?',
    connection_types: {
      wifi: 'Wi-Fi',
      wired: 'Wired Network',
      ether2ether: 'Direct Connection',
      usb: 'USB Connection',
    },
    connect_usb: {
      title_sub: ' (HEXA Only)',
    },
    back: 'back',
  },
}));

const mockWindowLocationReload = jest.fn();
jest.mock('app/actions/windowLocation', () => () => mockWindowLocationReload());

describe('test SelectConnectionType', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    window.location.hash = '?model=ado1';
  });

  test('should render correctly', () => {
    const { container, getByText } = render(<SelectConnectionType />);
    expect(container).toMatchSnapshot();

    fireEvent.click(getByText('Wi-Fi'));
    expect(window.location.hash).toBe('#initialize/connect/connect-wi-fi?model=ado1');

    fireEvent.click(getByText('Wired Network'));
    expect(window.location.hash).toBe('#initialize/connect/connect-wired?model=ado1');

    fireEvent.click(getByText('Direct Connection'));
    expect(window.location.hash).toBe('#initialize/connect/connect-ethernet?model=ado1');

    fireEvent.click(getByText('USB Connection'));
    expect(window.location.hash).toBe('#initialize/connect/connect-usb?model=ado1');

    fireEvent.click(getByText('back'));
    expect(window.location.hash).toBe('#initialize/connect/select-machine-model');
  });
});
