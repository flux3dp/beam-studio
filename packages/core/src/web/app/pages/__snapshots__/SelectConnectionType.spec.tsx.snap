// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SelectConnectionType should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="top-bar"
    />
    <div
      class="btns"
    >
      <div
        class="btn primary"
      >
        back
      </div>
    </div>
    <div
      class="main"
    >
      <h1
        class="title"
      >
        How do you wish to connect?
      </h1>
      <div
        class="btns"
      >
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wifi.svg"
          />
          <button
            class="btn btn"
            id="connect-wifi"
            type="button"
          >
            Wi-Fi
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-wired.svg"
          />
          <button
            class="btn btn"
            id="connect-wired"
            type="button"
          >
            Wired Network
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-ether2ether.svg"
          />
          <button
            class="btn btn"
            id="connect-ether2ether"
            type="button"
          >
            Direct Connection
          </button>
        </div>
        <div
          class="btn-container"
        >
          <img
            class="icon"
            draggable="false"
            src="img/init-panel/icon-usb.svg"
          />
          <button
            class="btn btn"
            id="connect-usb"
            type="button"
          >
            USB Connection
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
